### ISSUE TITLE
inserting timekeeping devices into a container doesn't enable the time display

### ISSUE THREAD

**Issue by shadowsapex** [CleverRaven/Cataclysm-DDA#68075]
### Describe the bug

when inserting into a container from the inventory menu, or moving between containers from the inventory organization menu, a timekeeping device won't enable its timekeeping function. if it was already in your inventory, and you insert it into another container, the time display will disappear. if you drop the item or the container and pick it back up, timekeeping will be reenabled.

the calorie counter function of a smartphone for example also gets disabled from this.

### Attach save file

[Plattsmouth-trimmed.tar.gz](https://github.com/CleverRaven/Cataclysm-DDA/files/12567835/Plattsmouth-trimmed.tar.gz)

### Steps to reproduce

spawn a smartphone and any container
open the container's inventory menu and use the insert option OR use the inventory organization screen to move the smartphone into the container
see that precise timekeeping is disabled (if no other timekeeping devices in inventory) and precise calorie counts are disabled in the eat menu

### Expected behavior

bug reverted to normal (i don't think i've ever had this problem in older versions)

### Versions and configuration

- OS: Windows
    - OS Version: 10.0.22621.2134 (22H2)
- Game Version: 73fd431 [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Portal Storms Ignore NPCs [personal_portal_storms],
    Slowdown Fungal Growth [no_fungal_growth]
]

**Comment by ehughsbaird** [comment]
Looks related to https://github.com/CleverRaven/Cataclysm-DDA/pull/67835

**Comment by Nebnis** [comment]
Looks like this is also affecting binoculars ability to see far away from inside backpacks

**Comment by ThePainkiller** [comment]
I can confirm this bug. Even wearing a wrist watch on the outermost layer of clothing makes one unable to tell the time on loading a save until you take it off and re-equip it. Smartphones inside pant's pockets won't work either even if not hidden, whether inside a phone case or not.

**Comment by ehughsbaird** [comment]
Probably this and https://github.com/CleverRaven/Cataclysm-DDA/issues/68435 are the same bug

**Comment by Lym666XD** [comment]
Good issue. I would be trying my best for this issue!

**Comment by Lym666XD** [comment]
Hi Guys, may I have a try on this issue?

**Comment by Kamayana** [comment]
I'm working on it, but I had to put it on hold to take care of another job. I should be able to get back to it soon.

**Comment by Lym666XD** [comment]
Thank you!  I've recently started working on the same issue for an academic assignment. I completely understand and respect the work you're putting into it. However, I was wondering if you might be able to grant me a window of around 8 days to work on and finalize my contributions for the assignment's sake. After this period, I'd be more than happy to collaborate, combine our findings, or step back to allow you to continue your work. I genuinely appreciate your understanding and cooperation.

**Comment by Lym666XD** [comment]
I'd like to apologize if my message comes across as too straightforward or aggressive. I'm not a native speaker, and sometimes I might not convey my intentions as delicately as I wish. I genuinely hope you're doing well.

I've recently started working on the same issue for an academic assignment (https://comp.anu.edu.au/courses/comp2120/assignments/assignment5/) after spending long time finding issues and am quite passionate about it. I completely respect the effort you're putting into this. Given the nature of my assignment, I was wondering if you might consider giving me a brief window of about 8 days to work on and finalize my contributions for the assignment. It would mean a lot to me. I would be really really gratefule for considering my request!

**Comment by Kamayana** [comment]
Yeah that's okay, go on ahead with your fix. My own branch isn't actually in a good state to show off, I mostly just determined that the core of the issue is that items added to the player's pockets via `insert_item_activity` need to be added to the character's caches with `Character.add_to_inv_search_caches`.

**Comment by Lym666XD** [comment]
Thank you!!!I am really grateful for your understanding and for sharing the insights about the issue.

**Comment by Lym666XD** [comment]
about the position to add `add_to_inv_search_caches` , I tried to add it in `insert_item_activity_actor::finish` function. But when using  `insert_item_activity_actor::finish`  to insert an item from ground to a container, `add_to_inv_search_caches` failed to insert the item to 'inv_search_caches'

**Comment by Kamayana** [comment]
That would be because the `holstered_item` reference is not the final item result - inside of `try_insert` the item is copied and that copy is placed inside the container (its pockets' `std::list<item> contents`), and that copy is now the ongoing memory location of the item. The original item gets deleted in `holstered_item.first.remove_item()`, so adding it to the cache just adds garbage data.

The problem is, as far as I can tell, there's currently no clean way to get a reference to the item copy that was created inside the container. One solution would be to change `try_insert` from `ret_val<void>` to `ret_val<item *>`, with `ret.value` being a pointer to the inserted item, but all of the insertion functions would have to also be modified, all the way down to `item_pocket::insert_item`, to feed a pointer up the chain to `try_insert`.

**Comment by andrei8l** [comment]
/confirmed The PR did not fix the issue. Kamayana's analysis above is correct.

**Comment by Kamayana** [comment]
I think I'm going to start working on this myself since I have an idea of how to fix it. I assume @Lym666XD's assignment is finished since it's been a few weeks now.

**Comment by Lym666XD** [comment]
Thanks guys , I am really sorry to failed on fixing the issue. I was taking exams these days and will be looking into the issues again soon.
